### DISCUSSION TITLE
Scalability concerns (high memory usage, long build times)

### DISCUSSION THREAD

**Discussion by krillboi** [facebook/docusaurus#3132] [General]
So I just finished deploying a new version of our docs (https://support.touchgfx.com/) which introduces versioning. It functions well but I noticed some concerning values when it comes to build times and memory usage. This particular version uses `docusaurus alpha.54`, so if these issues have been relieved in newer versions, I apologise (we will upgrade to newest Docuaurus version before our next release).

First of all, we have to use the `--max_old_space_size=16000` option on the node command to get it to finish, else it runs out of memory. It also took about 26 minutes to build the documentation as can be seen in the following image:

![docusaurus-build-times](https://user-images.githubusercontent.com/59022282/88541286-c9b18880-d014-11ea-94da-681bc589a3bd.png)

Furthermore, I saw upwards of 10gb of ram being used at some points during the build.

![docusaurus-memory-usage](https://user-images.githubusercontent.com/59022282/88541363-e5b52a00-d014-11ea-8732-8b0bdf911565.png)

For reference, this is building two versions of our docs (each with about 360 .mdx files) + the `next` version which we don't use (it seems I can disable building a `next` version in a newer Docusaurus version which is nice).

So it seems that adding versions is really tough for the build times and resources, so I thought I'd bring some awareness to this issue if you hadn't already considered it 😄 

Really appreciate the great work you guys are doing!

**Comment by slorber** [comment]
Hi,

How does it compare to the situation, you had before?

`--max_old_space_size=16000` is 16Gb that seems huge.

Have you tried lower values?

Have you tried with another CI, or building locally?

360 docs is already a large site, and everytime you'll cut a new version you'll increase the build time. Unfortunately we can't do much to avoid making the build time increase in such case. If this becomes a problem you can take older versions and publish them as standalone sites.

At some point if your site has many docs, many versions, many languages, yes you will have large build times without splitting the site in smaller subsites.

**Reply by krillboi** [reply]
The build is currently only being run locally. 

I just tried removing one version (which means it's building 1 version + next) and building. Memory usage didn't seem to change much. This is the result:

![image](https://user-images.githubusercontent.com/59022282/88547553-3715e700-d01e-11ea-9fa7-5a8d6800185d.png)

I also tried building while removing versioning. Memory usage dropped to a max of about 5gb. Result:

![image](https://user-images.githubusercontent.com/59022282/88548408-58c39e00-d01f-11ea-8a47-aa3317510ab6.png)

I realise that our site is large and that it will of course be unavoidable that build times will increase with every added version. I guess I was just surprised with how much each version seemed to increase build times/resource usage and there might be some optimization that could be done.

But yeah, I guess at some point having subsites for versions/languages could be a good idea.

**Reply by slorber** [reply]
I'm surprised too that it's so expensive to build your site, as it seems much slower than the Docusaurus 2 website (much less docs, but more versions, and build relatively fast)

If you give me the sources I could check.

Maybe you use too much the ideal image plugin? Image processing without caching has always been a bottleneck (including in Gatsby)

**Reply by krillboi** [reply]
Is the ideal image plugin something special? We usually use images with our own React components that uses a normal `<img>` tag.

Looking at our `static/img` folder, we currently have 544 images taking up 130mb of space.

I'll look into maybe providing you with the sources for the build.

**Reply by slorber** [reply]
The ideal image is not special, it's just an image transformation plugin, which is indeed a heavy process, that adds more build time everytime you add new ideal images

**Comment by krillboi** [comment]
@slorber I've gotten the go-ahead to provide you with the sources as long as I can send it to you personally. How should I send it to you?

**Reply by slorber** [reply]
if you have a protected link online you can send it to my personal email or via Discord

**Reply by csestito** [reply]
@krillboi @slorber were you ever able to find some optimizations to this issue?

**Reply by krillboi** [reply]
Unfortunately not, we have 4 versions now and a Chinese translation coming up, and building the whole site now takes approx. 2 hours and consumes over 17gb of RAM, so my concerns are still there 😄

**Reply by krillboi** [reply]
I think we're going to go for a multi-domain setup for each language in the future though, but just the versions alone take 50+ min to build.

**Reply by slorber** [reply]
@RDIL is working on the Webpack 5 migration here and it looks to be working: https://github.com/facebook/docusaurus/pull/4089

Trying to finish a few things before and catch up with some issues, but it is a high priority for me to do this upgrade and see how to improve the build time with some better caching.

Also want to provide a guide on how to scale a Docusaurus site, sharding it by locale, version or docs plugin instance. I think this is already possible but not too idiomatic.

**Reply by RDIL** [reply]
Hello yes I have arrived

The migration is almost ready, and should help out.

**Reply by slorber** [reply]
btw already contacted @sserrata on Discord as your site will be a nice case-study to run some benchmarks and we definitively want to improve your build times. Will tell you when it's a good time to work together on finding a solution

**Reply by sserrata** [reply]
FYI @glicht

**Comment by liaocyintl** [comment]
just try 
```
yarn build --no-minify
```
which reduce my building time from 30+m to 3m.

**Reply by slorber** [reply]
Surprising to see such a huge impact, are you sure it's not Webpack cache kicking in on second build (in addition to disabled modification)?

**Reply by liaocyintl** [reply]
> are you sure it's not Webpack cache kicking in on second build

I did `yarn clear` before `yarn build --no-minify`.
It works!

**Reply by slorber** [reply]
on our own website and my M1 computer `--no-minify` leads to 50s -> 45s, not as significant, so still surprised 😅 is your site opensource?

**Reply by liaocyintl** [reply]
> is your site opensource?

Unfortunately, it is an internal site.

> leads to 50s -> 45s

Maybe the 50s is not a significant long build time. If you generate a lot of MDXes on your site and increase the standard building time to more than 30m, --no-minify may be as significant.

**Reply by anaclumos** [reply]
```
    ~/Downloads/www  on   main !7 ?1  hyperfine --runs 20 --warmup 3 'npx docusaurus clear && npx docusaurus build --no-minify' 'npx docusaurus clear && npx docusaurus build'
Benchmark 1: npx docusaurus clear && npx docusaurus build --no-minify
  Time (mean ± σ):     337.329 s ±  2.618 s    [User: 501.452 s, System: 68.826 s]
  Range (min … max):   332.143 s … 342.655 s    20 runs

Benchmark 2: npx docusaurus clear && npx docusaurus build
  Time (mean ± σ):     430.421 s ±  3.707 s    [User: 880.611 s, System: 83.959 s]
  Range (min … max):   425.927 s … 438.454 s    20 runs

Summary
  'npx docusaurus clear && npx docusaurus build --no-minify' ran
    1.28 ± 0.01 times faster than 'npx docusaurus clear && npx docusaurus build'

✔  took 4h 57m 8s   at 03:00:01 PM 

```

I confirm that this makes the building 1.3x faster 😄

**Comment by OrKoN** [comment]
I think we encountered the same problem. We recently moved to Docusarus for [Puppeteer](https://github.com/puppeteer/puppeteer) and we try to generate a versioned documentation website. At first, the performance seemed good enough but as soon we started publishing more versions the process started to take more memory and time, to the point, when building the site using GitHub Actions fails with out-of-memory errors or takes many hours (not sure if it's hanging or if actually building the site). I wonder if there are workarounds to this problem? The source of the site is here: https://github.com/puppeteer/puppeteer/tree/main/website

**Reply by liaocyintl** [reply]
> GitHub Actions fails with out-of-memory errors

You can try to config the swap of Github Runner as follows.
It works for us.
ref. https://github.com/marketplace/actions/set-swap-space

```
- name: Set Swap Space
  uses: pierotofy/set-swap-space@master
  with:
    swap-size-gb: 10
```

**Reply by slorber** [reply]
@OrKoN once a version becomes unmaintained, I suggest "archiving it" so it's not anymore inside your production SPA and does not increase the build time.

We do this on the Docusaurus site itself:
- https://github.com/facebook/docusaurus/blob/main/website/versionsArchived.json
- https://github.com/facebook/docusaurus/blob/main/website/docusaurus.config.js

It's not an std feature but can be achieved in userland with a dynamic config.

You can also try to use swc (we use it on our website) or esbuild, see https://github.com/facebook/docusaurus/issues/4765

**Reply by OrKoN** [reply]
Thanks for the tips! We started to "archive" the old versions although it does not seem to be easy with GitHub pages so we just link to the markdown documentation for now. It'd be great if docusaurus could support light-weight versioning too when the previous version could be just moved to a sub-folder without re-building.

**Reply by slorber** [reply]
Yes @OrKoN , I'd like to provide a `docusaurus version:archive 1.0.0` CLI feature or something similar. It is a bit opinionated and possible to achieve in userland already today. I'd rather have more userland success stories of archiving documentation versions before designing this as a first-class feature: we want to implement the opinionated patterns that already work best for our user base.

**Comment by madelson** [comment]
I think we have a similar issue. We are trying to use Docusaurus to generate an internal site with both hand-written documentation files and auto-generated markdown files for API reference (something like the [MSFT docs site](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2?view=net-6.0)).

We have ~3100 markdown files in total, most of which are the auto-generated API reference docs. 

With this configuration, production build is _very_ slow (20+ minutes) and often runs out of memory.

**Reply by LunaticMuch** [reply]
I have a similar problem. Takes 60m on average.

**Comment by madelson** [comment]
@slorber / @RDIL any thoughts on next steps here? 
* Is the cause of the slowness known? Any suggestions on where to look in the docusaurus source?
* Are the maintainers interested in/planning to support larger sites with this tool?

**Reply by RDIL** [reply]
Hey @madelson,

Quick disclaimer: I'm not a Docusaurus maintainer, just an outside developer tooling engineer, so I can't speak to what the team will and won't do, but I can help relay your concerns to them.

The cause is known, and there is a fix available, but it comes with it's own drawbacks.  The causes, 2 of them actually, are:

### Cause 1: Babel

Babel is what is used to change the source code of both your site and Docusaurus's internals into what your browser can actually run.  Fancy things like React JSX don't work natively in a browser, instead they need to be rewritten by a tool (such as Babel) into somewhat more verbose and less readable code that will actually be run.

Babel is slow.  The reasoning behind this is because Babel is also a fully-fledged AST transformation kit (e.g. you can write custom plugins that allow you to generate new outputs, etc).  The big performance issue comes in when Babel has to transform a lot.  What's mainly taking so much time on very big sites is Node.js garbage collection - trying to free up extra memory that is no longer needed.  

How can this be fixed?  You can switch your JS loader to a faster alternative.  Many people recommend SWC and esbuild.  I'd personally go with esbuild, only because its documentation is more thorough.  How to do that can be found [here](https://github.com/facebook/docusaurus/issues/4765#issuecomment-870413073).  
Drawbacks: you can't use custom Babel plugins, and stuff may break (albeit, quite infrequently) if some strange edge case comes up.

### Cause 2: Terser

Terser makes the JS files smaller so that visitors aren't downloading tons of JS assets.  Terser is very well documented as being slow.

This one is the lesser of the 2 problems, but still exists, however a solution is much harder to come by for this issue.

**Reply by RDIL** [reply]
TL;DR: The cause is known, it's mainly Babel, but the maintainers don't want to switch off of it because it could break things for people that use custom Babel plugins/presets.  It should greatly help your site to switch from Babel to esbuild, which is actually relatively simple and easy to do.

**Reply by madelson** [reply]
Thanks @RDIL this is super helpful. For switching from Babel to ESBuild are there any instructions/examples you can point to or docusaurus-specific things to know going in?

WRT Terser, I'd happily trade off payload size for faster build times for my site; when you say a solution for this is "much harder to come by" can you explain a bit more?

We'll give the Babel switch a try and report back on what impact that makes.

**Reply by RDIL** [reply]
How to switch from Babel to esbuild in 3 steps:
1. Add `esbuild` and `esbuild-loader` to your dependencies in `package.json`
2. Add the following to the top level of your Docusaurus config:
```js
 webpack: {
    jsLoader: (isServer) => ({
      loader: require.resolve('esbuild-loader'),
      options: {
        loader: 'tsx',
        format: isServer ? 'cjs' : undefined,
        target: isServer ? 'node14' : 'es2017',
      },
    }),
  },
```
3. Experiment with it a bit - have somebody take a look through your site for maybe 5 to 10 minutes after making the switch (pretty sure nothing will break, unless you have things like custom components or embedded content from other websites).

For Terser, I think you can add the `--no-minify` flag to your build args, which should disable it.  The reason i'm hesitant though is that loading times can go up by like 20 seconds or more on slower networks such as mobile phones or computers on bad WiFi.  You *can* disable it, but it's really worth checking what the bundle size difference actually is - maybe also do a bit of benchmarking in Chrome DevTools with Lighthouse's performance measurements.  Just to reiterate, yes, it will speed things up, *but* at the cost of hurting users who don't have good connections.

**Reply by madelson** [reply]
Thanks @RDIL this was super helpful. I've tried this out on our site and it runs much more quickly (22 mins down to ~3 mins) but then it seems that esbuild crashes late in the process:

```
> node --max_old_space_size=3072 node_modules/@docusaurus/core/bin/docusaurus.mjs build

[INFO] [en] Creating an optimized production build...

● Client █████████████████████████ emitting (95%)
 emit

✖ Server
  Compiled with some errors in 2.13m

[ERROR] Docusaurus server-side rendering could not render static page with path /404.html.

Error: Cannot find module './assets/js/19743.182487ac.js'
Require stack:
- main
    at Array.reduce (<anonymous>)
    at Array.map (<anonymous>)

✔ Client
  Compiled successfully in 2.15m

✖ Server

● Client █████████████████████████ cache (99%) shutdown IdleFileCachePlugin
 resolve build dependencies

● Server █████████████████████████ cache (99%) shutdown IdleFileCachePlugin
 process pending cache items

[ERROR] Unable to build website for locale en.
[ERROR] Error: Failed to compile with errors.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! apt-platform-documentation@0.0.0 build: `node --max_old_space_size=3072 node_modules/@docusaurus/core/bin/docusaurus.mjs build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the my-documentation@0.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\madelson\AppData\Roaming\npm-cache\_logs\2022-09-09T13_45_37_554Z-debug.log
```

Contents of `C:\Users\madelson\AppData\Roaming\npm-cache\_logs\2022-09-09T13_45_37_554Z-debug.log`:
```
0 info it worked if it ends with ok
1 verbose cli [
1 verbose cli   'C:\\Program Files\\nodejs\\node.exe',
1 verbose cli   'C:\\Users\\madelson\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js',
1 verbose cli   'run',
1 verbose cli   'build'
1 verbose cli ]
2 info using npm@6.14.15
3 info using node@v16.14.2
4 verbose run-script [ 'prebuild', 'build', 'postbuild' ]
5 info lifecycle my-documentation@0.0.0~prebuild: my-documentation@0.0.0
6 info lifecycle my-documentation@0.0.0~build: my-documentation@0.0.0
7 verbose lifecycle my-documentation@0.0.0~build: unsafe-perm in lifecycle true
8 verbose lifecycle my-documentation@0.0.0~build: PATH: <omitted>
9 verbose lifecycle my-documentation@0.0.0~build: CWD: C:\dev\my-documentation\site
10 silly lifecycle my-documentation@0.0.0~build: Args: [
10 silly lifecycle   '/d /s /c',
10 silly lifecycle   'node --max_old_space_size=3072 node_modules/@docusaurus/core/bin/docusaurus.mjs build'
10 silly lifecycle ]
11 silly lifecycle my-documentation@0.0.0~build: Returned: code: 1  signal: null
12 info lifecycle my-documentation@0.0.0~build: Failed to exec build script
13 verbose stack Error: my-documentation@0.0.0 build: `node --max_old_space_size=3072 node_modules/@docusaurus/core/bin/docusaurus.mjs build`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (C:\Users\madelson\AppData\Roaming\npm\node_modules\npm\node_modules\npm-lifecycle\index.js:332:16)
13 verbose stack     at EventEmitter.emit (node:events:526:28)
13 verbose stack     at ChildProcess.<anonymous> (C:\Users\madelson\AppData\Roaming\npm\node_modules\npm\node_modules\npm-lifecycle\lib\spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (node:events:526:28)
13 verbose stack     at maybeClose (node:internal/child_process:1092:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:302:5)
14 verbose pkgid my-documentation@0.0.0
15 verbose cwd C:\dev\my-documentation\site
16 verbose Windows_NT 10.0.18363
17 verbose argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\madelson\\AppData\\Roaming\\npm\\node_modules\\npm\\bin\\npm-cli.js" "run" "build"
18 verbose node v16.14.2
19 verbose npm  v6.14.15
20 error code ELIFECYCLE
21 error errno 1
22 error my-documentation@0.0.0 build: `node --max_old_space_size=3072 node_modules/@docusaurus/core/bin/docusaurus.mjs build`
22 error Exit status 1
23 error Failed at the my-documentation@0.0.0 build script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

I'll continue to investigate but if you have any other insights to share from your own experience with these tools I'd welcome them!

**Reply by RDIL** [reply]
@madelson huh, that's a new one.  Are you able to share the repository with me for debugging purposes?  I can sign an NDA or something if you need me to.

**Reply by madelson** [reply]
@RDIL really appreciate the offer. I can't easily share our repo as-is but next week I'll try to reproduce with a "sanitized" version that could be shared.

**Reply by RDIL** [reply]
That's fine.  All I really need is `package.json`, `package-lock.json`, and your docusaurus config.

**Reply by Josh-Cena** [reply]
@madelson Remember to output CJS during server build (dynamic imports need to be compiled). Here's our config, before we migrated to SWC:

https://github.com/facebook/docusaurus/blob/74f653dd82d9db9f34dc28ecc0cf4340b4e309b1/website/docusaurus.config.js#L95-L103

It's `node12`, not `node14`. There's a very complicated reason behind this.

**Reply by madelson** [reply]
@Josh-Cena ok node14->node12 seemed to do the trick. I'd be very curious to understand the "very complicated reason" :-)

Here's what I have now:
```
    webpack: {
      jsLoader: (isServer) => ({
        loader: require.resolve('esbuild-loader'),
        options: {
          loader: 'tsx',
          format: isServer ? 'cjs' : undefined, // TODO do I need this?
          target: isServer ? 'node12' : 'es2017',
        },
      }),
    },
```
**Any thoughts on whether being explicit about cjs is needed and if so why?**

As far as performance, overall this is now clocking in at around **8m** overall:
* Client: 1.68m
* Server: 3.34m
* docusaurus-lunr-search <1 min
* "Generated static files in "build": ~2 min
I'm curious what all is happening in the last step and whether there is anything that can be done to optimize that, but still getting it down to 8 minutes from 22 minutes is a big win. I also haven't tried removing terser yet.

**Reply by RDIL** [reply]
I do believe that generating static files in build is either Terser doing its thing or a docusaurus post-processing task like link checking.

**Reply by RDIL** [reply]
![C1A2D173-8524-420F-BDD0-5964CD8E2024](https://user-images.githubusercontent.com/34555510/189693775-8c147af6-f0f2-48a6-be12-341df3f82179.jpeg)

Why it has to be Node 12 ^^

**Comment by LunaticMuch** [comment]
This morning on my CICD.

```
[2022-09-28T07:56:45.804Z] [INFO] [en] Creating an optimized production build...
[2022-09-28T08:35:40.264Z] [info] [webpackbar] Compiling Client
[2022-09-28T08:35:40.264Z] [info] [webpackbar] Compiling Server
[2022-09-28T08:37:47.266Z] [success] [webpackbar] Client: Compiled successfully in 2.35m
[2022-09-28T08:38:26.602Z] [success] [webpackbar] Server: Compiled successfully in 2.94m
[2022-09-28T08:38:53.114Z] [SUCCESS] Generated static files in "server/public".
```

Is there a way to get more details on steps executed during the build and their time?

**Comment by anaclumos** [comment]
More ways to speed up the build? i.e. build multiple locales in parallel?

**Reply by slorber** [reply]
This should be possible yes, but depends on your infra and is a bit out of the scope of Docusaurus:
- With GitHub CI you could use a [GitHub Actions Matrix](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs)
- With Netlify/Vercel/Others you could create one separate deployment per language. Using one subdomain per language, or assembling them under a single domain with rewrite rules.

**Reply by aybarsayan** [reply]
@slorber I'm facing a similar issue. I'm planning to host over 10,000 markdown files on the site, but it seems this might be impossible to manage it would probably ask for 60+ GB Ram. Using subdomains would be expensive for my use case. Would it be possible to implement dynamic page rendering instead? This would allow companies to create date-based powered websites more efficiently. For example, when a user requests a specific version's documentation, we could render it on-demand rather than building all versions during the build process. This could significantly reduce build times and resource usage while still maintaining versioning capabilities.

**Reply by slorber** [reply]
We don't plan to introduce a server runtime to do SSR for now.

Also keep in mind that what consumes time and memory is often bundling. Doing SSR still requires to bundle your app to create that server runtime.

Also have you tried Rspack? It consumes way less memory now
https://docusaurus.io/blog/releases/3.6

**Comment by tot-ra** [comment]
hitting same issue. [simple blog](https://github.com/tot-ra/kurapov.ee/), dies because it uses too much RAM when building docusaurus. I'm using pnpm btw.:

```
[INFO] [ru] Creating an optimized production build...
Killed
 ELIFECYCLE  Command failed with exit code 137.
Error: Process completed with exit code 137.
```

Tried rspack but it does not help
```
  future: {
    experimental_faster: {
      swcJsLoader: true,
      swcJsMinimizer: true,
      swcHtmlMinimizer: true,
      lightningCssMinimizer: true,
      rspackBundler: true,
      mdxCrossCompilerCache: true,
    },
  },
  
  ```
  
  As mitigation, forced to build it on my mac and push built html files to github :(

**Reply by slorber** [reply]
@tot-ra this discussion is quite old, please create a new one with all the relevant information. It worked before so you probably broke something.

https://github.com/tot-ra/kurapov.ee/commit/3f9e0ce5c424239bc8de1684d6846d907e47210c
This is now how we install Rspack, please read our documentation

`USE_EXPERIMENTAL_RSPACK=true`: this never existed

To get additional logs to see what step is failing, you can use `export DOCUSAURUS_PERF_LOGGER=true`

Please show me a proper code attempt, and give me a link to the associated CI logs showing the failure with that flag turned on

**Reply by anaclumos** [reply]
Hey @slorber, FYI I pruned old images from our Git history, which cut build times by roughly 3×. The image-compression script had been accumulating duplicates across commits, leading Docusaurus to hang during the “sealing asset processing” step. I disabled git history resolution anyways, perhaps it's doing unncessary works?

**Reply by slorber** [reply]
I don't understand what you mean. Does it refer any past discussion?

What script? What duplicates across commits? 

Regarding git history resolution (last update?), I just merged this PR: https://github.com/facebook/docusaurus/pull/11163

**Reply by anaclumos** [reply]
Oh, my apologies.

- I have a custom image compression script that runs automatically with every commit.
- This adds stress to the git repository (over 5 GB) because all images look the same, but with different sizes
- I believe the git history resolution module in Docusaurus takes quite some time to process images with extensive git history.
- Consequently, I pruned the images' history and found a threefold increase in build times.

**Reply by slorber** [reply]
What is the "Git history module resolution"? The git commands we execute to retrieve the last update time of an MDX file?

If yes, we only run this on the site load phase (which happens before bundling). It can take a decent amount of time (for large repos, it's often more costly than simply reading the actual MDX file content).

For this [11k docs site with 3k Git commits](https://github.com/sailpoint-oss/developer.sailpoint.com)
- With "last update disabled", the site loads in 7sec
- With "last update enabled", the site loads in 25 seconds (this is something I'll try to optimize in the future, but it's not significant compared to the bundling time)


---

> I believe the git history resolution module in Docusaurus takes quite some time to process images with extensive git history.

What makes you believe that?

I don't see how clearing your Git history would improve the bundling phase. Please create a dedicated issue with a repro so that I can investigate, because I'm really surprised. There's no place we read from Git while bundling.
